### ISSUE TITLE
Async requests are duplicated when sent without throttling

### ISSUE THREAD

**Issue by ilyazub** [serpapi/public-roadmap#688]
With @tanys123 while working on https://github.com/serpapi/SerpApi/issues/3426, we've found a race condition in the `/search` endpoint. Searches can be duplicated when sent without throttling in batches of 1k+ requests.

#### Send 1k requests via [cURL in parallel](https://stackoverflow.com/a/71967814/1291371)

```bash
curl -s --compressed --parallel 'https://serpapi.com/search?engine=google&q=race+condition&async=true&not_existing_dummy_parameter_for_curl_z=[1-1000]&api_key=$API_KEY'
```

#### Three searches were made instead of one: [1](https://serpapi.com/searches/640065692fe302a8ad79d79f/inspect), [2](https://serpapi.com/searches/640065692c084474f7d4a125/inspect), [3](https://serpapi.com/searches/6400653b1baebbc1a7f63c1a/inspect)

![image](https://user-images.githubusercontent.com/282605/222387009-24920ef5-1b5b-45e2-b5b6-ef5443ac7f5c.png)

### Why it's important

There are tons of users who make `async=true` requests. If someone will send 1k similar requests, several of them will be duplicated.

### Assumptions

This happens due to how `async=true` searches are processed in the background on SerpApi backend.

**Comment by hartator** [comment]
> This happens due to how `async=true` searches are processed in the background on SerpApi backend.

I would say sync searches should be even more prone to this because they return only `Success` searches? Ref: https://github.com/serpapi/SerpApi/blob/master/app/models/search.rb#L571

(Not sure super how to address, maybe with a blocking `Search.find_one_or_create` MongoDB call? But, maybe we are going to create more problems than we are solving.)

**Comment by hartator** [comment]
> Is it by design for sync searches?

Yes, I think so. To allow retries if a sync search is stuck in `Processsing`?

> Agree 👍 I'll wait for the feedback from customers (not sure if this edge case is noticeable for any of our customers).

Yes, I think so. If they manage to send them that many duplicate searches at the same time, it's on them?

The cache was meant more for serving already-successful searches than to avoid duplication.

**Comment by ilyazub** [comment]
Closing as [`wontfix`](https://github.com/serpapi/public-roadmap/labels/wontfix).
